packages/dds/tree/src/simple-tree/api/dirtyIndex.ts

```diff
@@ -0,0 +1,155 @@
+ * * "changed": A direct child was updated, added, removed, or moved (and the node is not "new").
+ * * "moved": The node was moved between/within arrays (and the node is not "new" or "changed").
+ * @alpha
+ */
+export type DirtyTreeStatus = "new" | "changed" | "moved";
+
+/**
+ * A map-like interface for tracking the {@link DirtyTreeStatus | status} of nodes that have been {@link trackDirtyNodes | tracked for changes}.
+ * @privateRemarks TODO: Replace this with `MapGetSet` from `@fluidframework/core-interfaces`.
+ * @alpha
+ */
+export interface DirtyTreeMap {
+	get(node: TreeNode): DirtyTreeStatus | undefined;
+	set(node: TreeNode, status: DirtyTreeStatus): void;
+}
+
+/**
+ * Registers a visitor on the view's forest that tracks which nodes are dirty.
+ * @param view - The view to track dirty nodes on
+ * @param dirty - A {@link DirtyTreeMap | map} that will be updated over time to reflect the {@link DirtyTreeStatus | status} of nodes as they change.
+ * Nodes that have not changed will not be inserted/updated in the map.
+ * @returns a cleanup function that should be called when the tracking is no longer needed.
+ * @example
+ * ```typescript
+ * const dirty = new Map<TreeNode, DirtyTreeStatus>();
+ * const stopTracking = trackDirtyNodes(view, dirty);
+ * // ... make changes to the view ...
+ * console.log(`The root of the tree is ${dirty.get(view.root) ?? "unchanged"}`);
+ * stopTracking();
+ * ```
+ * @alpha
+ */
+export function trackDirtyNodes(
+	view: TreeViewAlpha<ImplicitFieldSchema>,
+	dirty: DirtyTreeMap,
+): () => void {
+	const forest = (view as SchematizingSimpleTreeView<ImplicitFieldSchema>).checkout.forest;
+	const announcedVisitor = (): AnnouncedVisitor => createDirtyVisitor(forest, dirty);
+	forest.registerAnnouncedVisitor(announcedVisitor);
+	return () => {
+		forest.deregisterAnnouncedVisitor(announcedVisitor);
+	};
+}
+
+function createDirtyVisitor(forest: IEditableForest, dirty: DirtyTreeMap): AnnouncedVisitor {
+	// When cursor is in Fields mode, `parentField` is the field and `parent` is the parent node above that field (if any).
+	// When cursor is in Nodes mode, `parent` is the current node and `parentField` is undefined.
+	let parentField: FieldKey | undefined;
+	let parent: UpPath | undefined;
+
+	return createAnnouncedVisitor({
+		beforeDetach: (src) => {
+			assert(parent !== undefined, "Expected node");
+			assert(parentField !== undefined, "Expected field");
+			for (let parentIndex = src.start; parentIndex < src.end; parentIndex++) {
+				const path: UpPath = {
+					parent,
+					parentField,
+					parentIndex,
+				};
+				// The only way a detached node can be re-attached (and become visible/usable again) is via a move, so mark it as moved ahead of time.
+				const node = getNodeAtPath(forest, path);
+				if (node !== undefined && dirty.get(node) === undefined) {
+					// Only mark the node as moved if it is not already marked as something else.
+					dirty.set(node, "moved");
+				}
```

Comment:
It would be nice to have a helper for setting `DirtyTreeStatus` which handles the logic of status priority. Or change the `DirtyTreeMap` interface so that it's responsible for that internally.